src/core/remio.c

```diff
@@ -0,0 +1,668 @@
+        objpool_free(&remio_request_event_pool, event);
+        return ret;
+    }
+
+    /** Calculate the remaining number of pending I/O requests */
+    ret = (long int)remio_get_request_event_count(device);
+
+    remio_set_request_state(event->cpu_id, event->vcpu_id, REMIO_STATE_PROCESSING);
+
+    struct remio_request* request = remio_get_request(event->cpu_id, event->vcpu_id);
+
+    /** Write the I/O request information to the backend VM's vCPU registers */
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(0), request->addr);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(1), request->op);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(2), request->value);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(3), request->access_width);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(4), event->cpu_id);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(5), event->vcpu_id);
+
+    objpool_free(&remio_request_event_pool, event);
+
+    return ret;
+}
+
+/**
+ * @brief Handles the Remote I/O read and write operations
+ * @param value Value to be written or read
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remio_handle_rw(unsigned long value, unsigned long cpu_id, unsigned long vcpu_id)
+{
+    if (remio_get_request_state(cpu_id, vcpu_id) != REMIO_STATE_PROCESSING) {
+        return false;
+    }
+    remio_set_request_value(cpu_id, vcpu_id, value);
+    remio_set_request_state(cpu_id, vcpu_id, REMIO_STATE_COMPLETE);
+    return true;
+}
+
+/**
+ * @brief Performs the post work after the completion of the I/O request
+ * @note This function is executed by the frontend VM and is responsible for updating the
+ *       vCPU register in case of a read operation and activating the frontend vCPU
+ * @param event Message event (REMIO_CPU_MSG_*)
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remio_cpu_post_work(uint32_t event, uint8_t cpu_id, uint8_t vcpu_id)
+{
+    if (remio_get_request_state(cpu_id, vcpu_id) != REMIO_STATE_COMPLETE) {
+        return false;
+    }
+    struct remio_request* request = remio_get_request(cpu_id, vcpu_id);
+
+    switch (event) {
+        case REMIO_CPU_MSG_READ:
+            vcpu_writereg(cpu()->vcpu, request->reg, request->value);
+            break;
+        default:
+            break;
+    }
+
+    remio_set_request_state(cpu_id, vcpu_id, REMIO_STATE_FREE);
+
+    cpu()->vcpu->active = true;
+
+    return true;
+}
+
+long int remio_hypercall(unsigned long arg0, unsigned long arg1, unsigned long arg2)
+{
+    long int ret = -HC_E_SUCCESS;
+    unsigned long virt_remio_dev_id = arg0;
```

Comment:
why "virt"
should be remio_dev_id, or just dev_id

Reply:
The `virt` keyword implies that the backend VM is unaware of the absolute/remote I/O ID. For example, if there are two backend VMs, each with two remote I/O devices, the backend will work with devices IDs of 0 and 1. It is the responsibility of the Remote I/O system to calculate the correct absolute ID.

That said, I renamed `abs_remio_dev_id` to `remio_dev_id` for better clarity and alignment with its function.

Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/b79edf26d690526668c2a363f5f979e80fe1ee1e